### ISSUE TITLE
Request: Implement a workaround/compatibility solution for Motor City Online

### ISSUE THREAD

**Issue by SimplyTadpole** [AlpyneDreams/d8vk#121]
Hello! I wanted to apologize in advance if this isn't a proper usage of this project's issue reports - this technically isn't an issue with D8VK itself as it also occurs with vanilla Wine, and I couldn't find anything on this page related to requesting support for a specific game, but Motor City Online is a game that's mostly too obscure for upstream Wine developers to consider it a high-priority to fix at the moment (and understandably so; they have to focus on games and programs that are more popular rather than a dead game less than a dozen of people still play nowadays, lol), and since D8VK is a project actively focusing on older games, I hoped it would at least be okay to make a page here.

Motor City Online (I will refer to it as "MCO" from this point onwards) was an MMO from 2001 that was later shut down in 2003, and people made an unofficial launcher based on a leaked debug tool that allowed people to play it in a limited offline mode against AI opponents... unfortunately, for some reason, even this tool doesn't seem to let people play the game on Wine, as although the game will install successfully and the debug launcher will open, atttempting to actually launch the game will always cause the game to crash immediately, usually with the following error message:

![image](https://user-images.githubusercontent.com/70490001/225405262-2cf481ad-7e2e-4925-b3d7-c969099fa635.png)

This appears to happen regardless of the graphics card the host is actually using (I use a Radeon RX 570, which obviously has hardware acceleration, plus also discards the possibility of it being buggy Nvidia drivers). It's possible to bypass this error message through the Wine registry error by going to "HKLM\Software\Wow6432node\Electronic Arts\Motor City" and setting the value "D3D Device" to 0, but this merely leads to another issue - the game will launch and show the loading screen for a split-second, only to immediately crash anyway, with a different error message:

![image](https://user-images.githubusercontent.com/70490001/225407714-b07007af-3700-47a3-bf44-4ae12121a31d.png)

The errors, as well as the log files Wine accrued, seem to imply this issue may be related specifically to graphics rather than Wine itself, which hopefully means this could be fixed on D8VK's end without having to rely on upstream. On the other hand, a developer working on a rewrite of MCO's engine said that apparently the game utilizes "direct video card access", which "makes things like wine, virtualbox, [...] etc really unhappy". I don't exactly know what that entails, or if it can even be resolved on D8VK's end, but it sounds bad from my layperson perspective, and since said developer has been dealing with this program for 7 years, I'm inclined to believe they are speaking the truth on this.

I'm dearly hoping it's possible to add a workaround to this, though; I had a lot of fun playing this game when I was on Windows, and I'd be sad to lose the ability to play this game on Linux.

### Software information

• **Name of the game:** Motor City Online
• **Settings used:** I attempted everything - a 32-bit prefix; Windows version to XP, 98, 10; dgVoodoo; disabling DXVK; nGlide; regular Wine; Wine-GE; Wine-TKG; no matter what I tried, nothing would work. Not even D8VK seemed to do the trick - though it's worth noting that D8VK seems to only be available for 64-bit prefixes, and MCO will crash if it's not running on a 32-bit prefix apparently. (Well, at least crash harder than it does on a 32-bit prefix.)

### System information
- **GPU:** Radeon RX 570
- **Driver:** amdgpu
- **Wine version**: 8.3; 7.0; Wine-GE-Proton-7.35
- **D8VK version:** 0.10 - New Croak

### Apitrace file(s)
I couldn't really figure out how to get the Apitrace thingy to work as MCO uses an unique d3d8 file...

### Log files
- **d3d8.log:** None were generated as the game crashes immediately before a log could be generated
- **Wine log:** I uploaded two.
[mco log 1.txt](https://github.com/AlpyneDreams/d8vk/files/10983698/mco.log.1.txt)
[mco log 2.txt](https://github.com/AlpyneDreams/d8vk/files/10983699/mco.log.2.txt)

Again, terribly sorry if this isn't the proper place for this, or if I wrote this request wrong. If the latter, please let me know what I need to add.

**Comment by WinterSnowfall** [comment]
The good news is that the game is reportedly already working in Windows with d8vk. There's already some fixes for it in main, 8ddf18f IIRC, which should address problems with d3d rendering (it's actually mostly the game taking a very weird approach...).
<details>

<summary>See here</summary>

![image](https://user-images.githubusercontent.com/6306593/225433272-2ccb4263-3e22-4802-89a1-b8973be062d4.png)

</details>


The only thing remaining is to get it working in Wine, and see why it crashes.

**Comment by SimplyTadpole** [comment]
Is there anything I can do to help sort out the crash issue? But I'm guessing that means this is probably out of scope for D8VK? :c

**Comment by dungeon007** [comment]
This seems to be a problem with any D3D8... only combo that i found to make it work is d3d8 wrapper from Better Rayman:, run on DGVoodoo2's D3D9... and of course to DXVK DX11 or vkd3d-proton D3D12:

![BetterRayman+VKD3D](https://user-images.githubusercontent.com/49257090/225752095-a2fc7820-52e8-4832-a4cb-8b9b6fb50e86.png)

**Comment by SimplyTadpole** [comment]
Whoa whoa whoa, tell me more: is there a way to get the d3d8 wrapper from Better Rayman 3 without having to install Rayman 3 solely to run their setup installer? And how can I run DgVoodoo in DX9 mode? When I tried it, it seemed to only let me use it on DX11 mode :/

Also, since apparently Better Rayman 3's custom DX8 wrapper makes MCO run successfully if paired with dgVoodoo, maybe that's a good enough base for the D8VK team to figure out a fix?

**EDIT:** I attempted to follow your instructions, but no dice - MCO still crashes:

![image](https://user-images.githubusercontent.com/70490001/225940287-0007cf5f-b1a5-45f6-ba2f-ec0a37579814.png)
![image](https://user-images.githubusercontent.com/70490001/225940397-884db82a-2c1f-4ffe-bad1-f95ff7f1687f.png)
![image](https://user-images.githubusercontent.com/70490001/225940433-e7c84c6b-87ad-4559-ad1d-ff6930a3e81e.png)

This is the log file: [mco log 6.txt](https://github.com/AlpyneDreams/d8vk/files/11002631/mco.log.6.txt)

**Comment by SimplyTadpole** [comment]
I did the exact same steps, and it still crashes, with the exact same error messages I posted above.

Where did you get d3dx9_43.dll and d3dcompiler_43.dll, anyway? From winetricks? I don't have either of those in my installation folder :/

**Comment by SimplyTadpole** [comment]
...Okay, I managed to get it to work!! And the issue was incredibly stupid, too 😓

![image](https://user-images.githubusercontent.com/70490001/225982685-1e46f579-cb75-4533-a591-322e0e69fb11.png)

Following your instructions did the trick quite well (and installing Better Rayman 3 for the d3d8 wrapper didn't need Rayman 3 to be installed at all, thank goodness!), the issue was actually due to something completely unrelated - specifically, the Motor City update files have to replace a file called "text.eng", except the replacement file is called "text.ENG", and due to Linux's case-sensivity, the old file wasn't actually replaced, so the game kept trying to call the old one, which naturally made it crash. Deleting the old file and renaming the new one fixed it... I really should've considered doing this sooner :sweat_smile: 

Just for the sake of preservation, I'm uploading the necessary d3d8 files here in case someone wishes to play MCO on Linux and Better Rayman 3 ever gets taken down for whatever reason, [right here](https://cdn.discordapp.com/attachments/1053024581331464253/1086319764760187003/rayman_d3d8.7z).

Now, since it seems we've proven that it **is** possible to get Motor City Online to run on Linux with those bizarre nonsensical hacks, I think it's fair to ask that compatibility for it in D8VK itself be implemented, right? At least I'd like for others to be able to get this game running without having to resort to bizarre hacks involving another game entirely. :sweat_smile:

**Comment by dungeon007** [comment]
Well, it does not have to be wrapper from BR3, what you have there is basically identical d3d8.dll from dxwrapper version v1.0.990.18:
https://github.com/elishacloud/dxwrapper/releases
 It is notthing nonsencial, just older wrapper from year 2017 that happens to work here.
So probably one just have to bisecting around to see why newer does not :D

**Comment by WinterSnowfall** [comment]
I've got the game working with d8vk (in Wine), with just these steps:
a) "HKLM\Software\Wow6432node\Electronic Arts\Motor City" and setting the value "D3D Device" to 0
b) text.ENG -> text.eng when you apply the offline patch
c) [Setting the resolution to 800x600](https://www.pcgamingwiki.com/wiki/Motor_City_Online#Run_in_offline_mode)
d) make absolutely sure "D3D Device" is set to 0 again
e) Use the MCity_Launcher.exe to launch the game

![ApiTraceMode](https://user-images.githubusercontent.com/6306593/227802900-71ff70a3-e196-45fe-9564-bce74035939b.png)

PR #125 improves the abysmal performance and brings it to levels I think are on par with native. I doubt much more can be optimized for this game in particular, but things may get better once d8vk performance in general is further tweaked in the future.

**Comment by dungeon007** [comment]
Well, app is like horrybly CPU bound, probably because of that debug build being used... not much anybody could do there :D
I can run it on nine with wrapper, but with no updated dx8z.dll... just loading is horribly slow, like 2-3 minuted, again probably depends on a CPU.
 Also that dx8z.dll have broken fog... i found it is better to use VEGs wrapper from Need For Speed III Modern Patch https://veg.by/en/projects/nfs3/ beta build sould be OK, just rename dx8a.dll thrash driver to dx8z.dll and use it here with Motor City Offline.

**Comment by WinterSnowfall** [comment]
Just FYI, it's perfectly fine to try whatever combination gets the game launched for you, but simply by following the steps I have shared above, I can confirm it also just works with d8vk on Intel:

![MCOnline](https://user-images.githubusercontent.com/6306593/228382698-0e9afc7c-bb9e-4cdd-be87-08a582a6e36e.png)

Performance seems even better than it was on my RTX 3060, ironically. There's no artifacting or other problems of any nature really. The performance you're seeing is with PR #125 applied.

**Comment by dungeon007** [comment]
Well, that picture is obviosly low default settings on Intel, while on nvidia picture it wasnt on that... just look at a back window :D
 Perf diff that i am seeing, on d3d8to9 run on DGV2 from DEFAULT vs MAXED out is like about 3X. 
 
![default](https://user-images.githubusercontent.com/49257090/228428154-6413f883-47be-4043-979e-60f56b817676.png)

![maxed](https://user-images.githubusercontent.com/49257090/228428336-6d59f88f-9029-41a7-9eb0-4f92b43982c4.png)

**Comment by WinterSnowfall** [comment]
I did indeed take the screenshot on default settings, but then maxed them out and played the game for about an hour on Intel to compare performance. 

I believe you are correct in your assumption that the CPU is the bottleneck at this point, since I do have somewhat faster single-core performance on my laptop. Perhaps this situation was not clear when the game was designed (and GPUs were still insufficiently fast to expose this bottleneck), or they simply didn't know how to address it. 

In any case, although your fps counter will not launch into space with v-sync force disabled, the game does usually hold a very steady 60 fps even on my iGPU, so I believe we can consider this issue closed once my PR is merged. There's nothing we can magically do in d8vk if a game's engine is doing things suboptimally.

**Comment by Blisto91** [comment]
> Newer WINE (since 8.1) also defaults to Windows 10. So in future one can expect vsync to be broken on d3d8 apps, just like on Windows 10... so get ready for that, and dont be surprised :D

What do you mean? The Windows version in Wine is just some strings changed to spoof which Windows is installed afaik.

**Comment by dungeon007** [comment]
I mean, i expect WINE would break something there with vsync at some point just like on Windows 10.

**Comment by Blisto91** [comment]
Probably not. And if there is a issue relating to that it should be reported so it can be fixed or any specific games can have a workaround.
d8vk itself and drivers and etc etc also have a lot to say here.

**Comment by dungeon007** [comment]
Sure, just bewere of WINE 8.1+ and Windows 10 as default prefix... so lets hope not they wont break something in the same way.
 Meanwhile when you run dgVoodoo2 on vkd3d-proton, one already gets this feeling of new default. It just ignores what app sets there and expecting you to limit it :D

**Comment by Blisto91** [comment]
![image](https://user-images.githubusercontent.com/47954800/228952875-c6ee7dbd-4a41-4624-b84c-c97613400130.png)

Yay. Seems to work.
Gonna poke at it a bit.

**Comment by WinterSnowfall** [comment]
Right... so now that we have confirmation it works on all relevant GPU vendors, all is well in the world. Waiting on the PR merge to close the issue and thanks to @SimplyTadpole for bringing up this cool game many of us weren't even aware existed.

**Comment by drazisil** [comment]
It would be a lot cooler if you happen to know any C++ programmers who would be interested in pairing with me to get it back online. Which is it's normal state.

**Comment by dungeon007** [comment]
Yep, its normal state is also that one can inject any renderer into it... if somebody interested could make it online, he could make it load already existing GL renderers and if so desired likely to make VK renderer too. 
 But yes it is boring like this offline, one can beat AI easely even on hard. it is like a demo.

**Comment by drazisil** [comment]
I can get it online. I just need help. It's no longer a fun solo project after several years.

**Comment by Blisto91** [comment]
@drazisil sounds like a cool project. I don't personally know anyone like that but have linked your comment in a discord server.
